### ISSUE TITLE
how to use indexes when using Dali loader?

### ISSUE THREAD

**Issue by ludysama** [vturrisi/solo-learn#222]
I'm trying to select some samples dynamically while training, which need the sample's index to do it. 
I've achieved it using normal CPU loader, but it really cost a lot of time.

while reading the codes about Dali. I found an variable named 'conversion_map', Can I do some fixes on it, and then it returns the specific index of all the samples in my datasets? Is it available? How to make such a 'conversion_map'?

```
if self.conversion_map is not None:
            *all_X, indexes = [batch[v] for v in self.output_map]
            targets = self.conversion_map(indexes).flatten().long().detach().clone()
            indexes = indexes.flatten().long().detach().clone()
        else:
            *all_X, targets = [batch[v] for v in self.output_map]
            targets = targets.squeeze(-1).long().detach().clone()
            # creates dummy indexes
            indexes = (
                (
                    torch.arange(self.model_batch_size, device=self.model_device)
                    + (self.model_rank * self.model_batch_size)
                )
                .detach()
                .clone()
            )
```

**Comment by vturrisi** [comment]
Hey, @ludysama we already handle the indexes on our end. Dali does not support indexing images, so what we do is encode the index of the image into its label and then decode back the encoded index+label into index and label (which is handled in the conversion_map). To get the correct indexes of the images wrt to their order in the dataset, you just need to pass `--encode_indexes_into_labels` to your bash file as is done for DeepCluster. By default, the indexes that you get when using torchvision are the correct ones, but for Dali, if you don't use the flag, you will get sequential indexes from 0 to batch size*gpus. We don't return the actual indexes by default because there's a slight overhead in memory use.

We attribute the `conversion_map` in `solo/utils/dali_dataloader.py` if you wanna check how it works.

**Comment by ludysama** [comment]
thank you very much. It really helps a lot
